# Homework

**_NOTE:_** sometimes your answer doesn't match one of the options exactly. That's fine. Select the option that's closest to your solution.

## Dataset

In this homework, we will use the lead scoring dataset Bank Marketing dataset. Download it from here.

Or you can do it with wget:

```bash
wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv
```

In this dataset our desired target for classification task will be converted variable - has the client signed up to the platform or not.

## Data preparation

- Check if the missing values are presented in the features.
- If there are missing values:
  - For caterogiral features, replace them with 'NA'
  - For numerical features, replace with with 0.0

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

In [2]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv'

In [3]:
!wget $data -O course_lead_scoring.csv

--2025-10-08 21:08:59--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 80876 (79K) [text/plain]
Saving to: ‘course_lead_scoring.csv’

course_lead_scoring 100%[===================>]  78.98K  --.-KB/s    in 0.01s   

2025-10-08 21:09:00 (5.16 MB/s) - ‘course_lead_scoring.csv’ saved [80876/80876]



In [4]:
df = pd.read_csv('course_lead_scoring.csv')
df.head()

lead_source    industry  number_of_courses_viewed  annual_income  \
0      paid_ads         NaN                         1        79450.0   
1  social_media      retail                         1        46992.0   
2        events  healthcare                         5        78796.0   
3      paid_ads      retail                         2        83843.0   
4      referral   education                         3        85012.0   

  employment_status       location  interaction_count  lead_score  converted  
0        unemployed  south_america                  4        0.94          1  
1          employed  south_america                  1        0.80          0  
2        unemployed      australia                  3        0.69          1  
3               NaN      australia                  1        0.87          0  
4     self_employed         europe                  3        0.62          1

In [5]:
numerical = ['number_of_courses_viewed', 'annual_income', 'interaction_count', 'lead_score', 'converted']

In [6]:
df_num = df.select_dtypes(exclude='object')

In [7]:
df_num

number_of_courses_viewed  annual_income  interaction_count  lead_score  \
0                            1        79450.0                  4        0.94   
1                            1        46992.0                  1        0.80   
2                            5        78796.0                  3        0.69   
3                            2        83843.0                  1        0.87   
4                            3        85012.0                  3        0.62   
...                        ...            ...                ...         ...   
1457                         1            NaN                  4        0.53   
1458                         3        65259.0                  2        0.24   
1459                         1        45688.0                  3        0.02   
1460                         5        71016.0                  0        0.25   
1461                         3        92855.0                  3        0.41   

      converted  
0             1  
1             0  
2             1  
3             0  
4             1  
...         ...  
1457          1  
1458          1  
1459          1  
1460          1  
1461          1  

[1462 rows x 5 columns]

In [8]:
categorical = ['lead_source', 'industry', 'employment_status', 'location']

In [9]:
df_cat = df.select_dtypes(include=['object'])

In [10]:
df_cat

lead_source       industry employment_status       location
0           paid_ads            NaN        unemployed  south_america
1       social_media         retail          employed  south_america
2             events     healthcare        unemployed      australia
3           paid_ads         retail               NaN      australia
4           referral      education     self_employed         europe
...              ...            ...               ...            ...
1457        referral  manufacturing     self_employed  north_america
1458        referral     technology           student         europe
1459        paid_ads     technology           student  north_america
1460        referral            NaN     self_employed  north_america
1461  organic_search        finance           student  north_america

[1462 rows x 4 columns]

In [17]:
df_fill = df_num.fillna(0.0), df_cat.fillna('NA')

In [12]:
df_fill

(      number_of_courses_viewed  annual_income  interaction_count  lead_score  \
 0                            1        79450.0                  4        0.94   
 1                            1        46992.0                  1        0.80   
 2                            5        78796.0                  3        0.69   
 3                            2        83843.0                  1        0.87   
 4                            3        85012.0                  3        0.62   
 ...                        ...            ...                ...         ...   
 1457                         1            0.0                  4        0.53   
 1458                         3        65259.0                  2        0.24   
 1459                         1        45688.0                  3        0.02   
 1460                         5        71016.0                  0        0.25   
 1461                         3        92855.0                  3        0.41   
 
       converted  
 0     

# Question 1

What is the most frequent observation (mode) for the column industry?

- `NA`
- `technology`
- `healthcare`
- `retail`

In [38]:
df.industry.fillna('NA').mode()

0    retail
Name: industry, dtype: object

# Question 2

Create the correlation matrix for the numerical features of your dataset. In a correlation matrix, you compute the correlation coefficient between every pair of features.

What are the two features that have the biggest correlation?

- `interaction_count` and `lead_score`
- `number_of_courses_viewed` and `lead_score`
- `number_of_courses_viewed` and `interaction_count`
- `annual_income` and `interaction_count`

Only consider the pairs above when answering this question.
Split the data

- Split your data in train/val/test sets with 60%/20%/20% distribution.
- Use Scikit-Learn for that (the `train_test_split` function) and set the seed to ''42''.
- Make sure that the target value `y` is not in your dataframe.

# Question 3

- Calculate the mutual information score between y and other categorical variables in the dataset. Use the training set only.
- Round the scores to 2 decimals using round(score, 2).

Which of these variables has the biggest mutual information score?

- `industry`
- `location`
- `lead_source`
- `employment_status`

# Question 4

- Now let's train a logistic regression.
- Remember that we have several categorical variables in the dataset. Include them using one-hot encoding.
- Fit the model on the training dataset.
- To make sure the results are reproducible across different versions of Scikit-Learn, fit the model with these parameters:
- model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
- Calculate the accuracy on the validation dataset and round it to 2 decimal digits.

What accuracy did you get?

- 0.64
- 0.74
- 0.84
- 0.94

# Question 5

- Let's find the least useful feature using the feature elimination technique.
- Train a model using the same features and parameters as in Q4 (without rounding).
- Now exclude each feature from this set and train a model without it. Record the accuracy for each model.
- For each feature, calculate the difference between the original accuracy and the accuracy without the feature.

Which of following feature has the smallest difference?

- `'industry'`
- `'employment_status'`
- `'lead_score'`

**_NOTE:_** The difference doesn't have to be positive.

# Question 6

- Now let's train a regularized logistic regression.
- Let's try the following values of the parameter `C: [0.01, 0.1, 1, 10, 100]`.
- Train models using all the features as in Q4.
- Calculate the accuracy on the validation dataset and round it to 3 decimal digits.

Which of these `C` leads to the best accuracy on the validation set?

- 0.01
- 0.1
- 1
- 10
- 100

**_NOTE:_** If there are multiple options, select the smallest C.

# Submit the results

- Submit your results here: https://courses.datatalks.club/ml-zoomcamp-2025/homework/hw03
- If your answer doesn't match options exactly, select the closest one